master/internal/resourcemanagers/priority.go

```diff
@@ -0,0 +1,331 @@
+package resourcemanagers
+
+import (
+	"fmt"
+	"sort"
+
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/master/pkg/actor"
+	"github.com/determined-ai/determined/master/pkg/model"
+)
+
+type priorityScheduler struct {
+	preemptionEnabled bool
+}
+
+// NewPriorityScheduler creates a new scheduler that schedules tasks via priority.
+func NewPriorityScheduler(config *SchedulerConfig) Scheduler {
+	return &priorityScheduler{preemptionEnabled: config.Priority.Preemption}
+}
+
+func (p *priorityScheduler) Schedule(rp *ResourcePool) ([]*AllocateRequest, []*actor.Ref) {
+	return p.prioritySchedule(rp.taskList, rp.groups, rp.agents, rp.fittingMethod)
+}
+
+func (p *priorityScheduler) prioritySchedule(
+	taskList *taskList,
+	groups map[*actor.Ref]*group,
+	agents map[*actor.Ref]*agentState,
+	fittingMethod SoftConstraint,
+) ([]*AllocateRequest, []*actor.Ref) {
+	agentsSplitByLabel := splitAgentsByLabel(agents)
+	toAllocate := make([]*AllocateRequest, 0)
+	toRelease := make([]*actor.Ref, 0)
+
+	// Since labels are a hard scheduling constraint, process every
+	// label independently.
+	for label, agentsWithLabel := range agentsSplitByLabel {
```

Comment:
**question**: Is this to make sure a task isn't split across agents of different labels?
**blocking**: could this create a situation where a task ends up not being scheduled where it should be? example if you have task 1 (4 slots) and 2 (2 slots) and agents A(2 slot) and B(4 slot) - will this only scheduler task 1 on B?

Reply:
in our scheduler `agentLabels` is a hard constraints so if a task doesn't have a specified label we do not allow to be scheduled on an agent with that label and vice-versa. In effect, agent labels create independent clusters within a resource pool (they are mostly redundant with resource pools so in the future we likely will remove them). 

Since labels create independent clusters, we split them up here to perform scheduling independently on each one of them.